File: hash_table/go/hash.go
package hash_table

func hash(s string, mod int) int {
	return int(rawHash(s) % uint32(mod))
}

func rawHash(s string) uint32 {
	var h uint32
	for _, r := range s {
		h = 101*h + uint32(r)
	}
	return h
}

File: hash_table/javascript/hash.js
export default function hash(str, modulo) {
    const h = rawHash(str) % modulo;
    return h < 0 ? h + modulo : h;
}

function rawHash(str) {
    let h = 0;

    for (let i = 0; i < str.length; i += 1) {
        h = 101 * h + str.charCodeAt(i);

        // Bitwise arithmetic forces the result to be a 32-bit integer (see ECMAScript spec).
        // Other than that, h |= 0 (shorthand for h = h | 0) is a no-op.
        h |= 0;
    }

    return h;
}
